From a7e7e92f9c446ee2dfbbecf592211564b60ade4c Mon Sep 17 00:00:00 2001
From: Todd Whiteman <todd.whiteman@joyent.com>
Date: Tue, 31 Jul 2018 16:38:20 -0700
Subject: [PATCH] TRITON-630 cns and cmon content-md5 header incompatible with
 vmapi for non-ascii characters Reviewed by: Trent Mick <trentm@gmail.com>

---
 CHANGES.md                 |  4 ++++
 bin/cnsadm                 | 10 +++++-----
 lib/api-server.js          | 10 ++++++----
 lib/changefeed-filter.js   |  6 +++---
 lib/cn-filter.js           |  6 +++---
 lib/napi-legacy-filter.js  |  6 +++---
 lib/net-pool-filter.js     |  6 +++---
 lib/network-info-filter.js |  6 +++---
 lib/poller-stream.js       |  6 +++---
 lib/reaper-stream.js       |  6 +++---
 lib/utils.js               | 25 +++++++++++++++++++++++--
 package.json               |  4 ++--
 12 files changed, 61 insertions(+), 34 deletions(-)

diff --git a/CHANGES.md b/CHANGES.md
index fc845c8..35da7a2 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -10,6 +10,10 @@
 
 # triton-cns
 
+## 0.1.3
+
+- TRITON-630 cns and cmon content-md5 header incompatible with vmapi for non-ascii characters
+
 ## 0.1.2
 
 - TRITON-578 CNS should not continue as though it succeeded in the case of getNapiPools failure
diff --git a/bin/cnsadm b/bin/cnsadm
index 70c6122..c1e0a56 100755
--- a/bin/cnsadm
+++ b/bin/cnsadm
@@ -4,7 +4,7 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
  *
- * Copyright (c) 2015, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 var dashdash = require('dashdash');
@@ -108,9 +108,9 @@ if (Buffer.isBuffer(sapiUrl))
 	sapiUrl = sapiUrl.toString('ascii');
 sapiUrl = sapiUrl.trim();
 
-var sapi = restify.createJsonClient({
+var sapi = restify.createJsonClient(utils.getRestifyClientOptions({
 	url: sapiUrl
-});
+}));
 
 var SCHEMA = JSON.parse(
     fs.readFileSync('/opt/triton/cns/lib/config-schema.json').
@@ -158,9 +158,9 @@ sapi.get('/configs/' + zoneName, function (err, req, res, obj) {
 		process.exit(2);
 	}
 
-	apiClient = restify.createJsonClient({
+	apiClient = restify.createJsonClient(utils.getRestifyClientOptions({
 		url: 'http://127.0.0.1'
-	});
+	}));
 
 	sapi.get('/services?name=cns', function (err2, req2, res2, objs) {
 		if (err2) {
diff --git a/lib/api-server.js b/lib/api-server.js
index bc8f6c0..b9b1887 100644
--- a/lib/api-server.js
+++ b/lib/api-server.js
@@ -19,6 +19,7 @@ var net = require('net');
 var dns = require('dns');
 var consts = require('./consts');
 var restify = require('restify');
+var restifyClients = require('restify-clients');
 var child_process = require('child_process');
 var fs = require('fs');
 var cueball = require('cueball');
@@ -171,10 +172,11 @@ function APIServer(opts) {
 		return (ev);
 	};
 
-	this.napiClient = restify.createJsonClient({
-		url: 'http://' + this.config.napi_opts.address,
-		agent: agent
-	});
+	this.napiClient = restifyClients.createJsonClient(
+		utils.getRestifyClientOptions({
+			url: 'http://' + this.config.napi_opts.address,
+			agent: agent
+		}));
 }
 
 function tryLookup(addr, retries, timeout) {
diff --git a/lib/changefeed-filter.js b/lib/changefeed-filter.js
index bda8c69..8c576ac 100644
--- a/lib/changefeed-filter.js
+++ b/lib/changefeed-filter.js
@@ -3,7 +3,7 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
  *
- * Copyright (c) 2015, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 module.exports = ChangefeedFilter;
@@ -32,10 +32,10 @@ function ChangefeedFilter(opts) {
 
 	assert.optionalObject(opts.agent, 'options.agent');
 
-	this.client = restify.createJsonClient({
+	this.client = restify.createJsonClient(utils.getRestifyClientOptions({
 		url: 'http://' + this.config.address,
 		agent: opts.agent
-	});
+	}));
 
 	var streamOpts = {
 		readableObjectMode: true,
diff --git a/lib/cn-filter.js b/lib/cn-filter.js
index 7716bd5..b24671a 100644
--- a/lib/cn-filter.js
+++ b/lib/cn-filter.js
@@ -3,7 +3,7 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
  *
- * Copyright (c) 2015, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 module.exports = CNFilter;
@@ -38,10 +38,10 @@ function CNFilter(opts) {
 
 	assert.optionalObject(opts.agent, 'options.agent');
 
-	this.client = restify.createJsonClient({
+	this.client = restify.createJsonClient(utils.getRestifyClientOptions({
 		url: 'http://' + this.config.address,
 		agent: opts.agent
-	});
+	}));
 
 	this.cache = undefined;
 
diff --git a/lib/napi-legacy-filter.js b/lib/napi-legacy-filter.js
index 6dd3c65..6a8a981 100644
--- a/lib/napi-legacy-filter.js
+++ b/lib/napi-legacy-filter.js
@@ -3,7 +3,7 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
  *
- * Copyright (c) 2016, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 module.exports = NAPILegacyFilter;
@@ -34,10 +34,10 @@ function NAPILegacyFilter(opts) {
 
 	assert.optionalObject(opts.agent, 'options.agent');
 
-	this.client = restify.createJsonClient({
+	this.client = restify.createJsonClient(utils.getRestifyClientOptions({
 		url: 'http://' + this.config.address,
 		agent: opts.agent
-	});
+	}));
 
 	this.cache = LRUCache({
 		max: 32*1024*1024,
diff --git a/lib/net-pool-filter.js b/lib/net-pool-filter.js
index 5efd0e3..a5c3166 100644
--- a/lib/net-pool-filter.js
+++ b/lib/net-pool-filter.js
@@ -3,7 +3,7 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
  *
- * Copyright (c) 2015, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 module.exports = NetPoolFilter;
@@ -35,10 +35,10 @@ function NetPoolFilter(opts) {
 
 	assert.optionalObject(opts.agent, 'options.agent');
 
-	this.client = restify.createJsonClient({
+	this.client = restify.createJsonClient(utils.getRestifyClientOptions({
 		url: 'http://' + this.config.address,
 		agent: opts.agent
-	});
+	}));
 	this.cache = undefined;
 
 	this.timer = setInterval(this.updatePools.bind(this), UPDATE_INTERVAL);
diff --git a/lib/network-info-filter.js b/lib/network-info-filter.js
index 4a9e8de..ba2aba4 100644
--- a/lib/network-info-filter.js
+++ b/lib/network-info-filter.js
@@ -3,7 +3,7 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
  *
- * Copyright (c) 2016, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 module.exports = NetworkInfoFilter;
@@ -34,10 +34,10 @@ function NetworkInfoFilter(opts) {
 
 	assert.optionalObject(opts.agent, 'options.agent');
 
-	this.client = restify.createJsonClient({
+	this.client = restify.createJsonClient(utils.getRestifyClientOptions({
 		url: 'http://' + this.config.address,
 		agent: opts.agent
-	});
+	}));
 
 	this.cache = LRUCache({
 		max: 32 * 1024 * 1024,
diff --git a/lib/poller-stream.js b/lib/poller-stream.js
index 69a1001..f5313b9 100644
--- a/lib/poller-stream.js
+++ b/lib/poller-stream.js
@@ -3,7 +3,7 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
  *
- * Copyright (c) 2015, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 module.exports = PollerStream;
@@ -47,10 +47,10 @@ function PollerStream(opts) {
 
 	assert.optionalObject(opts.agent, 'options.agent');
 
-	this.client = restify.createJsonClient({
+	this.client = restify.createJsonClient(utils.getRestifyClientOptions({
 		url: 'http://' + this.config.address,
 		agent: opts.agent
-	});
+	}));
 
 	var streamOpts = {
 		objectMode: true,
diff --git a/lib/reaper-stream.js b/lib/reaper-stream.js
index 40bbbb3..e87c25f 100644
--- a/lib/reaper-stream.js
+++ b/lib/reaper-stream.js
@@ -3,7 +3,7 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
  *
- * Copyright (c) 2015, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 module.exports = ReaperStream;
@@ -51,10 +51,10 @@ function ReaperFSM(strm, opts) {
 	this.maxSleep = 10000;
 	this.reapTime = DEFAULT_REAP_TIME;
 
-	this.client = restify.createJsonClient({
+	this.client = restify.createJsonClient(utils.getRestifyClientOptions({
 		url: 'http://' + opts.config.vmapi_opts.address,
 		agent: opts.agent
-	});
+	}));
 
 	FSM.call(this, 'idle');
 }
diff --git a/lib/utils.js b/lib/utils.js
index df74888..e3aced4 100644
--- a/lib/utils.js
+++ b/lib/utils.js
@@ -3,7 +3,7 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
  *
- * Copyright (c) 2015, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 module.exports = {
@@ -13,10 +13,12 @@ module.exports = {
 	recSetMatch: recSetMatch,
 	recSetDiff: recSetDiff,
 	recMatch: recMatch,
-	cleanVM: cleanVM
+	cleanVM: cleanVM,
+	getRestifyClientOptions: getRestifyClientOptions
 };
 
 var ipaddr = require('ipaddr.js');
+var jsprim = require('jsprim');
 
 function currentSerial() {
 	var now = new Date();
@@ -173,3 +175,22 @@ function cleanVM(obj) {
 	obj.timers = [];
 	obj.timers.push({t: new Date(), n: 'cleanVM'});
 }
+
+/*
+ * Merge the given restify client options with a set of default restify options.
+ *
+ * @param {Object} opts - restify client options.
+ * @returns New object containing the merged restify client options.
+ */
+function getRestifyClientOptions(opts) {
+	var newOpts = jsprim.deepCopy(opts);
+
+	// TRITON-630 When calculating the response content-md5 hash, use both
+	// 'utf8' and 'binary' encodings in order to support node v4 (and
+	// before) talking to node v6 (and higher).
+	if (!newOpts.contentMd5) {
+		newOpts.contentMd5 = {encodings: ['utf8', 'binary']};
+	}
+
+	return (newOpts);
+}
diff --git a/package.json b/package.json
index 9921cb7..e63b5fc 100644
--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "triton-cns",
-  "version": "0.1.2",
+  "version": "0.1.3",
   "description": "Trition container naming service",
   "main": "index.js",
   "scripts": {
@@ -37,7 +37,7 @@
     "mname": "1.3.5",
     "redis": "2.1.0",
     "restify": "4.3.0",
-    "restify-clients": "1.5.0",
+    "restify-clients": "1.6.0",
     "restify-errors": "4.3.0",
     "sdc-scripts": "git+https://github.com/joyent/sdc-scripts.git#deefaef587ed3bee2706cb6e53ee3468e682932e",
     "sprintf-js": "1.0.3",
-- 
2.21.0

